### ISSUE TITLE
[C#] Implement new Concept Exercise: casting

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1142]
This issue describes how to implement the `casting` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the Concept of Casting in C#.

## Learning objectives

- Know what explicit and implicit casts are.
- Know how to do an explicit cast.
- Know how to use `is` and `as` to cast types.
- Know how to use the `typeof` operator.

## Out of scope

- Custom explicit and implicit cast operators.
- Pattern matching on types.
- Memory and performance characteristics.

## Concepts

This Concepts Exercise's Concepts are:

- `casting`: know what explicit and implicit casts are; know how to do an explicit cast; know how to use `is` and `as` to cast types.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `numbers`: know how to cast numbers and what explicit and implicit numeric casts are.
- `inheritance`: work with inheritance to show when casting can be reasonably used.
- `exceptions`: know about exceptions to understand what happens when an explicit cast fails.

### Hints

- [Type testing and cast operators][type-testing-and-cast-operators]: introduction to type testing and casting.
- [cast operator][cast-operator]: the cast operator.
- [`is` operator][is-operator]: `is` operator reference.
- [`as`-operator][as-operator]: `as` operator reference.
- [`typeof` operator][typeof-operator]: `typeof` operator reference.
- [Type conversion exceptions][type-conversion-exceptions]: explains when a runtime exception is thrown when doing casts.

### After

- [Type testing and cast operators][type-testing-and-cast-operators]: introduction to type testing and casting.
- [cast operator][cast-operator]: the cast operator.
- [`is` operator][is-operator]: `is` operator reference.
- [`as`-operator][as-operator]: `as` operator reference.
- [`typeof` operator][typeof-operator]: `typeof` operator reference.
- [Type conversion exceptions][type-conversion-exceptions]: explains when a runtime exception is thrown when doing casts.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/concepts/type_casting.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/–docs/concept-exercises.md
[type-testing-and-cast-operators]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast
[is-operator]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#is-operator
[as-operator]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#as-operator
[cast-operator]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#cast-operator-
[typeof-operator]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#typeof-operator
[type-conversion-exceptions]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/casting-and-type-conversions#type-conversion-exceptions-at-run-time

**Comment by mikedamay** [comment]
Consider declaring casts of primitive types out-of-scope and handling them as a concept in `signed-integers`.

**Comment by mikedamay** [comment]
closed by PR #1860
